csrc/scheduler/multi_matmul.cpp

```diff
@@ -65,6 +83,31 @@ void MultipleMatmulScheduler::findRoles() {
 
   as_ = tensor_roles_.at(MatmulTensorRole::OPERAND_A);
   bs_ = tensor_roles_.at(MatmulTensorRole::OPERAND_B);
+  // When translating MatmulOp or LinearOp with avoid_intermediates, we
+  // introduce some intermediate Global tensors which will be ignored during
+  // lowering. We update as_ and bs_ to point at the last of these tensors
+  // before their next consumer is in local memory.
+  for (std::vector<TensorView*>* tensors : {&as_, &bs_}) {
```

Comment:
I have a strong preference against `while (true)` loops.

```suggestion
    auto find_first_local_consumer = [](TensorView* tv) -> TensorView* {
    while (tv != nullptr) {
      NVF_ERROR(tv->uses().size() == 1);

      Expr* use = tv->uses().front();
      NVF_ERROR(use->outputs().size() == 1);

      NVF_ERROR(
          use->isA<BroadcastOp>() || use->isA<SqueezeOp>() ||
          use->isA<LoadStoreOp>());

      TensorView* consumer = dynamic_cast<TensorView*>(use->output(0));
      NVF_ERROR(consumer != nullptr);

      if (consumer->getMemoryType() == MemoryType::Local) {
        return tv;
      }

      // Traverse down consumers
      tv = consumer;
    }
    NVF_THROW("Failed to find a consumer that is a local tensor.");
    return nullptr;
  };
  
  // Apply in-place transformation
  std::transform(
      as_.cbegin(), as_.cend(), as_.begin(), find_first_local_consumer);
  std::transform(
      bs_.cbegin(), bs_.cend(), bs_.begin(), find_first_local_consumer);
```

Reply:
I made the `break` scenarios into `NVF_ERROR`. It seems like undefined behavior if we encounter an non-skippable global intermediate.

Reply:
I'll add a more helpful comment here. What we're doing is chasing the chain of replacements. For example, let's say we start with
```
Inputs:
  tv0
  tv1

tv2 = matmul(tv0, tv1)
```
After replacement this might become something like the following:
```
Inputs:
  tv0
  tv1

tv3 = broadcast(tv0)
tv4 = broadcast(tv1)
tv5 = permute(tv4)
tv2 = matmul(tv3, tv5)
```
What we are doing here is redefining `tensor_roles_[A]` from `{tv0}` to `{tv3}` and `tensor_roles_[B]` from `{tv1}` to `{tv5}`.

To do this, we are just advancing each of the original inputs as long as their consumer: 1) is unique 2) has a single use (in case of multi-matmul) and 3) still lies in Global memory.

So most of these really are stopping conditions and not unexpected. For example a multi-matmul Fusion with two matmuls that shares the same A operand would have a gmem operand with multiple uses.